public/app/features/dashboard-scene/panel-edit/PanelEditor.tsx

```diff
@@ -79,10 +90,63 @@ export class PanelEditor extends SceneObjectBase<PanelEditorState> {
       dashboard.onEnterEditMode();
     }
 
-    const panelMngr = this.state.panelRef.resolve();
+    const panelManager = this.state.panelRef.resolve();
 
-    if (sourcePanel!.parent instanceof SceneGridItem) {
-      sourcePanel!.parent.setState({ body: panelMngr.state.panel.clone() });
+    const sourcePanelParent = sourcePanel!.parent;
+    const sourcePanelGrandparent = sourcePanelParent!.parent!;
+    if (!(sourcePanelGrandparent instanceof SceneGridLayout)) {
+      console.error('Expected grandparent to be SceneGridLayout!');
+      return;
+    }
+
+    const normalToRepeat = !this._initialRepeatOptions.repeat && panelManager.state.repeat;
+    const repeatToNormal = this._initialRepeatOptions.repeat && !panelManager.state.repeat;
+
+    if (sourcePanelParent instanceof SceneGridItem) {
+      if (normalToRepeat) {
+        const repeatDirection = panelManager.state.repeatDirection ?? 'h';
+        const repeater = new PanelRepeaterGridItem({
+          key: sourcePanelParent.state.key,
+          x: sourcePanelParent.state.x,
+          y: sourcePanelParent.state.y,
+          width: repeatDirection === 'h' ? 24 : sourcePanelParent.state.width,
+          height: sourcePanelParent.state.height,
+          itemHeight: sourcePanelParent.state.height,
+          source: panelManager.state.panel.clone(),
+          variableName: panelManager.state.repeat!,
+          repeatedPanels: [],
+          repeatDirection: panelManager.state.repeatDirection,
+          maxPerRow: panelManager.state.maxPerRow,
+        });
+        sourcePanelGrandparent.setState({
+          children: sourcePanelGrandparent.state.children.map((child) =>
+            child.state.key === sourcePanelParent.state.key ? repeater : child
+          ),
+        });
+      } else {
+        sourcePanelParent.setState({ body: panelManager.state.panel.clone() });
+      }
+    } else if (sourcePanelParent instanceof PanelRepeaterGridItem) {
+      if (repeatToNormal) {
+        const panelClone = panelManager.state.panel.clone();
+        const gridItem = new SceneGridItem({
+          key: sourcePanelParent.state.key,
+          x: sourcePanelParent.state.x,
+          y: sourcePanelParent.state.y,
+          width: this._initialRepeatOptions.repeatDirection === 'h' ? 8 : sourcePanelParent.state.width,
+          height: this._initialRepeatOptions.repeatDirection === 'v' ? 8 : sourcePanelParent.state.height,
+          body: panelClone,
+        });
+        sourcePanelGrandparent.setState({
+          children: sourcePanelGrandparent.state.children.map((child) =>
+            child.state.key === sourcePanelParent.state.key ? gridItem : child
+          ),
+        });
+      } else {
+        sourcePanelParent.setState({ source: panelManager.state.panel.clone() });
+      }
+    } else {
+      console.error('Unsupported scene object type');
```

Comment:
Can we extract this to a separate method and also add some unit tests for this behavior? Quite a lot going on here, would be worth having some coverage.